src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,383 @@ public class SqlMutationEngine : IMutationEngine
-                // derive from TransactionException. Hence, TransactionException acts as a catch-all.
-                // When an exception related to Transactions is encountered, the mutation is deemed unsuccessful and
-                // a DataApiBuilderException is thrown
-                catch (TransactionException)
+                if (context.OperationType is EntityActionOperation.Upsert || context.OperationType is EntityActionOperation.UpsertIncremental)
                 {
-                    throw _dabExceptionWithTransactionErrorMessage;
-                }
+                    DbResultSet? upsertOperationResult;
+                    DbResultSetRow upsertOperationResultSetRow;
+                    JsonDocument? selectOperationResponse = null;
 
-                DbResultSetRow? dbResultSetRow = upsertOperationResult is not null ?
-                    (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
+                    // This variable indicates whether the upsert resulted in an update operation. If true, then the upsert resulted in an update operation.
+                    // If false, the upsert resulted in an insert operation.
+                    bool hasPerformedUpdate = false;
 
-                if (upsertOperationResult is not null &&
-                    dbResultSetRow is not null && dbResultSetRow.Columns.Count > 0)
-                {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
+                    try
+                    {
+                        // Creating an implicit transaction
+                        using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType(sqlMetadataProvider))
+                        {
+                            upsertOperationResult = await PerformUpsertOperation(
+                                                                parameters: parameters,
+                                                                context: context,
+                                                                sqlMetadataProvider: sqlMetadataProvider);
+
+                            if (upsertOperationResult is null)
+                            {
+                                // Ideally this case should not happen, however may occur due to unexpected reasons,
+                                // like the DbDataReader being null. We throw an exception
+                                // which will be returned as an Unexpected InternalServerError
+                                throw new DataApiBuilderException(
+                                    message: "An unexpected error occurred while trying to execute the query.",
+                                    statusCode: HttpStatusCode.InternalServerError,
+                                    subStatusCode: context.OperationType is EntityActionOperation.Upsert ? DataApiBuilderException.SubStatusCodes.UnexpectedRestUpsertOperationFailure
```

Comment:
The different substatuscodes seem to be an overkill here. Since ideally this case should not be happening, why does specifying an internal server error not seem sufficient?